cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -74,88 +83,160 @@
+     * @param value the value to be added
+     * @return a new {@link MetadataEntity.Builder} which is of same type of this {@link MetadataEntity.Builder}
+     * but consists of the given key and value in addition
+     */
+    public Builder append(String key, String value) {
+      parts.put(key.toLowerCase(), value);
+      return this;
+    }
+
+    /**
+     * Put the given key with the given value in {@link MetadataEntity.Builder}.
+     * The returned {@link MetadataEntity} type is set to the given key. If an append is required without the
+     * type change then {@link #append(String, String)} should be used.
+     *
+     * @param key the key to be added
+     * @param value the value to be added
+     * @return a new {@link MetadataEntity.Builder} whose type is set to the given key and consists of the given
+     * key and value in addition
+     */
+    public Builder appendAsType(String key, String value) {
+      parts.put(key.toLowerCase(), value);
+      this.type = key.toLowerCase();
+      return this;
+    }
+
+    /**
+     * Builds a {@link MetadataEntity} from the builder.
+     *
+     * @return {@link MetadataEntity} from the builder
+     * @throws IllegalArgumentException if the key is a CDAP entity and the MetadataEntity is not correct to represent
+     * the CDAP entity
+     */
+    public MetadataEntity build() {
+      if (parts.isEmpty() || type == null || type.isEmpty()) {
+        throw new IllegalArgumentException("key-value pair and type must be specified");
+      }
+      validateHierarchy();
+      return new MetadataEntity(Collections.unmodifiableMap(parts), type);
+    }
+
+    private void validateHierarchy() {
+      if (TYPES_TO_KEY_SEQUENCES.containsKey(type)) {
+        String[][] validSequences = TYPES_TO_KEY_SEQUENCES.get(type);
+        for (String[] validSequence : validSequences) {
+          if (Arrays.equals(validSequence, parts.keySet().toArray())) {
+            // valid sequence found
+            return;
+          }
+        }
+        throw new IllegalArgumentException(String.format("Failed to build MetadataEntity of type '%s' from '%s'. " +
+                                                           "Type '%s is a CDAP entity type and must follow one of " +
+                                                           "the following key hierarchies '%s'." +
+                                                           "If you want to represent a CDAP Entity please follow the " +
+                                                           "correct hierarchy. If you are trying to represent a " +
+                                                           "custom resource please use a different type name.",
+                                                         type, parts, type, Arrays.toString(validSequences)));
+      }
+    }
+  }
+
+  public MetadataEntity(MetadataEntity metadataEntity) {
+    this.details = new LinkedHashMap<>(metadataEntity.details);
```

Comment:
because details is already unmodifiable, no need to copy.

Reply:
Fixed.